### PR TITLE
Add task to specify which node version the CI should use

### PR DESCRIPTION
**Pull request by andre4i**
## Description

In order to not install node for every build (or every other build, depending on how caching works) we need to bake the required node version in the image (separate PR in a separate repository) and use `n` to manage the node version.

This should be merged only after the machine images have been updated with the required package (`n`).

### CODE REVIEW

File: tools/pipelines/templates/include-use-node-version.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+- name: version
+  type: string
+
+steps:
+- task: Bash@3
+  name: SetNodeVersion
+  displayName: Set node version
+  inputs:
+    targetType: 'inline'
+    workingDirectory: ${{ parameters.buildDirectory }}
```
[the remark this answers is not in the document]
Comment (andre4i) [reply tools/pipelines/templates/include-use-node-version.yml:16]: Good catch

**Comment by andre4i** [comment]
This will pass all tests as soon as I switch the image in the agent pools.

**Comment by andre4i** [comment]
> What's the plan for updating the node version after this change? Will the build fail if we updated the version used in the CI pipelines, but not the build vm images? Is there a way to fall back to installing the requested version if it's not in the image?

Hey @tylerbutler , the image template is in the internal repo which we can't link from here. The template can be configured to install whatever versions we need. I don't believe it's a good idea to fall back in case of a version mismatch. If we explicitly request a node version in pipeline definition and it's not supported by the image, we will know about it really fast. We can also preemptively install newer versions in the image definitions if that's the path we want to take. I will tag you in the PR dealing with the image template.

**Comment by andre4i** [comment]
✅  across the board

**Review by tyler-cai-microsoft — APPROVED** [review]
:shipit:

**Comment by tylerbutler** [comment]
I don't have any objections to this change, but I don't understand what we would have to update to use a new node version after this is merged.

For comparison, updating our pnpm version requires editing package.json for the release groups roots and that's all. (It's moderately secure because we include the hash, and node verifies that downloaded bits match the hash.) Since we know node will release a new minor or patch version at some point, how do we consume that new version? Is it the same as this change, but with a different version number? Something else?

Different scenario where the same question is relevant: I want to test the pipelines against a new version of node (let's say node 20) on a test branch. How do I do it?

**Comment by andre4i** [comment]
> I don't have any objections to this change, but I don't understand what we would have to update to use a new node version after this is merged.

> For comparison, updating our pnpm version requires editing package.json for the release groups roots and that's all. (It's moderately secure because we include the hash, and node verifies that downloaded bits match the hash.) Since we know node will release a new minor or patch version at some point, how do we consume that new version? Is it the same as this change, but with a different version number? Something else?

@tylerbutler 

You should have gotten an ADO notification from the PR in ff_internal which deals with the image. You can configure the image to install whatever node versions you want with `n` inside agent-image.json (look for `nodeVersions`). Currently the image has two node versions baked in, 16.16.0 and 18.16.1. They can be toggled by changing the parameters of the build task added with this PR. IF you need other versions of node, add them to the image template, rebuild the image and update the Azure resource group's Large/Small pools to point to the newly built image.

> I want to test the pipelines against a new version of node (let's say node 20) on a test branch. How do I do it?

Update the image to install node 20, rebuild the image, swap in the image and afterwards configure the CI to use 20 in your test branch.

**Review by alexvy86 — APPROVED** [review]

**Comment by tylerbutler** [comment]
Is there any item tracking making this easier? How does an open source contributor update the image? Why do we need to create a new image every single time there's a node release? This is very expensive, and the solution we had before made it trivial to use new versions. This seems like a major maintenance step backwards.

**Pull request merged**